addons/purchase_stock/tests/test_purchase_order_process.py

```diff
@@ -57,3 +59,40 @@ def test_01_packaging_propagation(self):
         })
         po.button_confirm()
         self.assertEqual(po.order_line.move_ids.product_packaging_id, packaging)
+
+    def test_02_vendor_delay_report_partially_cancelled_purchase_order(self):
+        """ Test vendor delay reports for partially cancelled purchase order"""
+        partner = self.partner_1
+        purchase_order = self.env['purchase.order'].create({
+            'partner_id': partner.id,
+            'date_planned': fields.Datetime.now().replace(hour=23, minute=0) + timedelta(days=2),
+            'order_line': [
+                Command.create({
+                    'product_id': self.product_1.id,
+                    'product_qty': 2.0,
+                    'product_uom': self.product_1.uom_id.id,
+                }),
+                Command.create({
+                    'product_id': self.product_2.id,
+                    'product_qty': 3.0,
+                    'product_uom': self.product_2.uom_id.id,
+                })],
+        })
+        purchase_order.button_confirm()
+        purchase_order.picking_ids.move_ids.flush_recordset()
+        delay_reports = self.env['vendor.delay.report'].read_group(fields=["__count", "on_time_rate:sum", "qty_total:sum"], domain=[('partner_id', "=", partner.id)], groupby=["product_id"])
+        self.assertEqual([rec['on_time_rate'] for rec in delay_reports], [0.0, 0.0])
+        # cancel the first part of the PO
+        purchase_order.order_line.filtered(lambda l: l.product_id == self.product_1).product_qty = 0
+        self.assertEqual(partner.purchase_order_count, 1)
+        self.assertTrue(float_compare(partner.on_time_rate, 0.0, precision_rounding=0.01) <= 0)
+        self.assertEqual(purchase_order.picking_ids.move_ids.filtered(lambda l: l.product_id == self.product_1).state, 'cancel')
+        purchase_order.picking_ids.move_ids.filtered(lambda l: l.product_id == self.product_2).quantity_done = 3.0
+        purchase_order.picking_ids.button_validate()
+        self.assertEqual(purchase_order.picking_ids.move_ids.filtered(lambda l: l.product_id == self.product_2).state, 'done')
+        self.assertEqual(partner.purchase_line_ids.mapped('qty_received'), [0.0, 3.0])
+        partner.invalidate_recordset(fnames=['on_time_rate'])
+        self.assertEqual(float_compare(partner.on_time_rate, 100.0, precision_rounding=0.01), 0.0)
```

Comment:
in other cases you cases you can use "assertAlmostEqual" but here a direct assert is fine
``` py
self.assertEqual(partner.on_time_rate, 100.0)
```